### ISSUE TITLE
Mocking methods with scalar type hinting and optional parameters fails with PHP 7.1

### ISSUE THREAD

**Issue by letr0n** [padraic/mockery#623]
Hi,

Mocking methods that use scalar type hints will fail when using PHP 7.1 if the parameter is optional:

```
public function foo(string $foo = null)
{

}
```

It produces a `ParseError`:

```
ParseError: syntax error, unexpected '?', expecting identifier (T_STRING)
```

This is because [ReflectionParameter::getType()](https://github.com/padraic/mockery/blob/master/library/Mockery/Generator/Parameter.php#L77) prepends the type with a question mark if the parameter is nullable in PHP 7.1.

I fixed it by changing [MethodDefinitionPass.php line 114](https://github.com/padraic/mockery/blob/master/library/Mockery/Generator/StringManipulation/Pass/MethodDefinitionPass.php#L114) from 

```
$typeHint = trim($param->getTypeHintAsString());
```

to

```
$typeHint = ltrim(trim($param->getTypeHintAsString()), '?');
```

**Comment by davedevelopment** [comment]
Thanks! I'll tag this issue and make sure it gets dealt with when I get on to the PHP 7.1 stuff.

**Comment by davedevelopment** [comment]
Actually, I don't think I have the rights to create labels, so I'll just need to remember it...

**Comment by Majkl578** [comment]
Any ETA for 7.1 support? It's RC3 already and this is a blocker for testing.

**Comment by davedevelopment** [comment]
@Majkl578 sorry, I haven't had chance to work on it yet.

**Comment by davedevelopment** [comment]
@letr0n Could you clarify your exact problem for me? I get the feeling your sample code is missing something...

``` php
<?php

require "vendor/autoload.php";

class Foo {
    public function foo(string $foo = null)
    { }
}

$mock  = Mockery::mock('Foo');

var_dump(phpversion()); // string(8) "7.1.0RC3"
var_dump($mock instanceof Foo); // bool(true)
```

**Comment by letr0n** [comment]
Sorry I forgot to update the issue. I belive it got reverted in PHP7.1 RC3

https://github.com/php/php-src/commit/bf907b9961000dfb9b3bf3bd8a1e00e327a8268e

**Comment by Majkl578** [comment]
RC3+ is needed for properly working type reflection, earlier versions were a mess... I guess this could be closed and new issue for 7.1 support should be opened. :)
Have a look at doctrine/common#734 for reference.
